java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/CheckerFrameworkUtil.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util;
+
+import org.checkerframework.dataflow.qual.AssertMethod;
+
+/**
+ * Utility methods for handling global configuration into CheckerFramework library.
+ */
+public class CheckerFrameworkUtil {
+
+  /**
+   * Method annotation @AssertMethod indicates that a method checks a value and possibly
+   * throws an assertion. Using it can make flow-sensitive type refinement more effective.
+   *
+   * @param eval The assertion to be evaluated at compile time
+   */
+  @AssertMethod
+  public static void assertMethod(boolean eval) {
```

Comment:
Alternatively, I'm led to believe that just bare `assert` is also understood by Checker Framework.

Reply:
> We already have Preconditions; can we just annotate that?

Uhmm, there are not relation between `Precondition` and CheckerFramework. There are validation at compile time that is evaluated by CheckerFramework, different than Preconditions that is evaluated at runtime.

Error:  `[dereference.of.nullable] dereference of possibly-null reference oldLedger`
````
final BufferLedger oldLedger = map.remove(allocator);
BufferAllocator oldAllocator = oldLedger.getAllocator();
````

Error:  `[dereference.of.nullable] dereference of possibly-null reference oldLedger`
````
final BufferLedger oldLedger = map.remove(allocator);
Preconditions.checkNotNull(oldLedger);
BufferAllocator oldAllocator = oldLedger.getAllocator();
````

No error: using @AssertMethod checker framework
````
final BufferLedger oldLedger = map.remove(allocator);
CheckerFrameworkUtil.assertMethod(oldLedger != null);
BufferAllocator oldAllocator = oldLedger.getAllocator();
````

Reply:
> Alternatively, I'm led to believe that just bare `assert` is also understood by Checker Framework.

Then, Do you agree to continue using `@AssertMethod` as it is implemented on this PR or do you recommend other `assert` option (please explain)?

Reply:
Right, but the point of the annotation to tell Checker Framework that Preconditions is an assertion. The point is to use it on your existing assertion methods.

Reply:
That said it appears to not support checkNotNull style things. However, does `assert oldLedger != null;` work?

Reply:
Alternatively, does it understand Objects.requireNonNull? 

What I want to avoid is having to add more redundant assertion helpers. We should annotate existing ones where possible.

Reply:
Let me try

Reply:
We have the following options:

1st. Introduce `@AssertMethod`:
````
final BufferLedger oldLedger = map.remove(allocator);
CheckerFrameworkUtil.assertMethod(oldLedger != null);
BufferAllocator oldAllocator = oldLedger.getAllocator();
````

2nd. Customize [assert](https://checkerframework.org/manual/#assumeassertion:~:text=32.2%E2%80%83%40-,AssumeAssertion%20string%20in%20an%20assert%20message,-Sometimes%2C%20it%20is) with `@AssumeAssertion(checkername)`
````
final BufferLedger oldLedger = map.remove(allocator);
assert oldLedger != null : "@AssumeAssertion(nullness): BufferLedger(oldLedger) must not be null";
BufferAllocator oldAllocator = oldLedger.getAllocator();
````

3rd. Objects.requireNonNull is out of scope because we assign Nullable as an option and then we are required to be NonNull:
````
final BufferLedger oldLedger = map.remove(allocator);
Objects.requireNonNull(oldLedger, "BufferLedger(oldLedger) must not be null");
BufferAllocator oldAllocator = oldLedger.getAllocator();
...
error: [argument] incompatible argument for parameter obj of Objects.requireNonNull.
[ERROR]   found   : @Initialized @Nullable BufferLedger
[ERROR]   required: @Initialized @NonNull BufferLedger

````

Could you please let me know if we will proceed with the first option (as it is) or move to the second option?

Reply:
Again: why can't we annotate https://github.com/apache/arrow/blob/48f704e2a3161366661d180e0c2198c00671756c/java/memory/memory-core/src/main/java/org/apache/arrow/util/Preconditions.java#L134-L138

Reply:
Ok, just added.